File: core/src/output_queue.rs
#[derive(Default)]
pub struct OutputQueue(String);
impl OutputQueue {
    pub fn new() -> Self {
        Self(String::new())
    }
    pub fn insert(&mut self, string: &str) -> Vec<String> {
        let queue = &mut self.0;
        queue.push_str(string);
        let mut vec = Vec::new();
        let mut i = 0;
        while let Some(pos) = queue[i..].find('\n').map(|i| i + 1) {
            vec.push(queue[i..pos].trim_end().to_string());
            i = pos;
        }
        self.0 = queue[i..].to_string();
        vec
    }
    pub fn left(&mut self) -> String {
        let string = self.0.to_string();
        self.0 = String::new();
        string
    }
}

File: playground/src/output_queue.ts
export class OutputQueue {
    private queue = "";
    insert(output: string): string[] {
        let { queue } = this;
        queue += output;
        let arr: string[] = [];
        let i = 0;
        while (true) {
            let pos = queue.indexOf("\n", i);
            if (pos < 0) break;
            pos++;
            arr.push(queue.slice(i, pos).trimEnd());
            i = pos;
        }
        this.queue = queue.slice(i);
        return arr;
    }
    left(): string {
        let { queue } = this;
        this.queue = "";
        return queue;
    }
}
